File: leetcode/math/1088_Confusing_Number_II/codes/math.py
class Solution:
    def confusingNumberII(self, n: int) -> int:
        confusing_digits = [0, 1, 6, 8, 9]
        mapping = {"0": "0", "1": "1", "6": "9", "8": "8", "9": "6"}
        s = str(n)
        half = len(s) // 2
        odd = bool(len(s) % 2)
        ans = 2
        for l in range(2, len(s)):
            ans += 4 * (5 ** (l - 1) - 5 ** (l // 2 - 1) * (3 if l % 2 == 1 else 1))

        def calc(idx: int) -> int:
            num = int(s[idx])
            if num > 8:
                index = 4
            elif num > 6:
                index = 3
            elif num > 1:
                index = 2
            elif num == 1:
                index = 1
            else:
                index = 0
            if idx == len(s) - 1:
                return index + (confusing_digits[index] == int(s[idx]))
            smaller_num = index - 1 if idx == 0 else index
            res = smaller_num * 5 ** (len(s) - idx - 1)
            if idx < half:
                res -= smaller_num * 5 ** (half - idx - 1) * (3 if odd else 1)
            if idx == half and odd:
                res -= 3 if num > 8 else (2 if num > 1 else (1 if num > 0 else 0))
            if confusing_digits[index] == int(s[idx]):
                res += calc(idx + 1)
            return res

        ans += calc(0)
        first_half_num = 0
        for i in range(half - 1, -1, -1):
            if s[i] not in mapping:
                return ans
            first_half_num = first_half_num * 10 + int(mapping[s[i]])
        if odd:
            if s[half] in {"0", "1", "8"} and int(s[half + 1:]) >= first_half_num:
                ans -= 1
        elif int(s[half:]) >= first_half_num:
                ans -= 1
        return ans

File: leetcode/math/1088_Confusing_Number_II/codes/math.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int confusingNumberII(int n) {
        vector<int> digits = {0, 1, 6, 8, 9};
        unordered_map<char, char> mapping = {
            {'0', '0'}, {'1', '1'}, {'6', '9'}, {'8', '8'}, {'9', '6'}
        };
        string s = to_string(n);
        int half = s.size() / 2;
        bool odd = s.size() % 2;
        int ans = 2;

        for (int l = 2; l < s.size(); l++) {
            int tmp = pow(5, l - 1);
            int sub = pow(5, l / 2 - 1) * (l % 2 ? 3 : 1);
            ans += 4 * (tmp - sub);
        }

        ans += calc(s, 0, digits, mapping, half, odd);

        int first_half_num = 0;
        for (int i = half - 1; i >= 0; i--) {
            if (!mapping.count(s[i])) return ans;
            first_half_num = first_half_num * 10 + (mapping[s[i]] - '0');
        }

        if (odd) {
            if (s[half] == '0' || s[half] == '1' || s[half] == '8') {
                if (stoi(s.substr(half + 1)) >= first_half_num) ans--;
            }
        } else {
            if (stoi(s.substr(half)) >= first_half_num) ans--;
        }
        return ans;
    }

    int calc(const string &s, int idx, const vector<int>& digits, unordered_map<char, char> &mapping, int half, bool odd) {
        int num = s[idx] - '0';
        int index = (num > 8) ? 4 : (num > 6) ? 3 : (num > 1) ? 2 : (num == 1) ? 1 : 0;

        if (idx == s.size() - 1) return index + (digits[index] == num);

        int smaller = (idx == 0 ? index - 1 : index);
        int res = smaller * pow(5, s.size() - idx - 1);

        if (idx < half)
            res -= smaller * pow(5, half - idx - 1) * (odd ? 3 : 1);
        if (idx == half && odd)
            res -= (num > 8 ? 3 : (num > 1 ? 2 : (num > 0 ? 1 : 0)));
        if (digits[index] == num)
            res += calc(s, idx + 1, digits, mapping, half, odd);
        return res;
    }
};
